c/xudt_rce.c

```diff
@@ -0,0 +1,481 @@
+  // 0 is the only success code. We can use 0 directly.
+
+  // inherit from simple_udt
+  ERROR_ARGUMENTS_LEN = -1,
+  ERROR_ENCODING = -2,
+  ERROR_SYSCALL = -3,
+  ERROR_SCRIPT_TOO_LONG = -21,
+  ERROR_OVERFLOWING = -51,
+  ERROR_AMOUNT = -52,
+
+  // error code is starting from 40, to avoid conflict with
+  // common error code in other scripts.
+  ERROR_CANT_LOAD_LIB = 40,
+  ERROR_INVALID_ARGS,
+  ERROR_NOT_ENOUGH_BUFF,
+  ERROR_INVALID_FLAG,
+  ERROR_INVALID_ARGS_FORMAT,
+  ERROR_INVALID_WITNESS_FORMAT,
+  ERROR_INVALID_MOL_FORMAT,
+  ERROR_BLAKE2B_ERROR,
+  ERROR_HASH_MISMATCHED,
+  ERROR_RCRULES_TOO_DEEP,
+  ERROR_TOO_MANY_RCRULES,
+  ERROR_RCRULES_PROOFS_MISMATCHED,
+  ERROR_SMT_VERIFY_FAILED,
+};
+
+#define CHECK2(cond, code) \
+  do {                     \
+    if (!(cond)) {         \
+      err = code;          \
+      ASSERT(0);           \
+      goto exit;           \
+    }                      \
+  } while (0)
+
+#define CHECK(code)  \
+  do {               \
+    if (code != 0) { \
+      err = code;    \
+      ASSERT(0);     \
+      goto exit;     \
+    }                \
+  } while (0)
+
+#define BLAKE2B_BLOCK_SIZE 32
+#define BLAKE160_SIZE 20
+#define SCRIPT_SIZE 32768
+#define WITNESS_SIZE 32768
+#define EXPORTED_FUNC_NAME "validate"
+#define MAX_CODE_SIZE (1024 * 1024)
+#define FLAGS_SIZE 4
+
+#include "rce.h"
+
+// global variables, type definitions, etc
+
+// We will leverage gcc's 128-bit integer extension here for number crunching.
+typedef unsigned __int128 uint128_t;
+
+uint8_t g_script[SCRIPT_SIZE] = {0};
+uint8_t g_witness[WITNESS_SIZE] = {0};
+uint32_t g_witness_len = 0;
+bool g_witness_inited = false;
+
+uint8_t g_code_buffer[MAX_CODE_SIZE] __attribute__((aligned(RISCV_PGSIZE)));
+uint32_t g_code_used = 0;
+
+/*
+is_owner_mode indicates if current xUDT is unlocked via owner mode(as
+described by sUDT), extension_index refers to the index of current extension in
+the ScriptVec structure. args and args_length are set to the script args
+included in Script structure of current extension script.
+
+If this function returns 0, the validation for current extension script is
+consider successful.
+ */
+typedef int (*ValidateFuncType)(int is_owner_mode, size_t extension_index,
+                                const uint8_t* args, size_t args_len);
+
+typedef enum XUDTFlags {
+  XUDTFlags_Plain = 0,
+  XUDTFlags_InArgs = 1,
+  XUDTFlags_InWitness = 2,
+} XUDTFlags;
+
+typedef enum XUDTValidateFuncCategory {
+  XVFC_Normal = 0,  // normal extension script
+  XVFC_RCE = 1,     // Regulation Compliance Extension
+} XUDTValidateFuncCategory;
+
+uint8_t RCE_HASH[32] = {1};
+
+// functions
+int load_validate_func(const uint8_t* hash, uint8_t hash_type,
+                       ValidateFuncType* func, XUDTValidateFuncCategory* cat) {
+  int err = 0;
+  void* handle = NULL;
+  size_t consumed_size = 0;
+
+  if (memcmp(RCE_HASH, hash, 32) == 0) {
+    *cat = XVFC_RCE;
+    *func = rce_validate;
+    return 0;
+  }
```

Comment:
Let's add a new mode here, and make the validation flow like this:

* If the extension script indicates a pre-defined one (such as RCE), we will use hardcoded function here
* If the extension script is identical to a lock script of one input cell in current transaction, we consider the extension script to be already validated, no additional check is needed for current extension
* As a final resort, we will try the dynamic linking path

Reply:
add "is_extension_script_validated" for step 2.